### ISSUE TITLE
Investigate and implement new git workflow to support 4.x

### ISSUE THREAD

**Issue by pyrello** [uiowa/uids#671]
# Stories
- As a developer, I can maintain the existing version of UIDS while developing a new version.
- As a developer, I can consume multiple versions of UIDS in my application.

# Proposed solution
Investigation to determine how to set up a git workflow to allow development of multiple versions of UIDS simultaneously. This could be similar to how Drupal module developers maintain multiple versions of their modules.

## Questions to resolve
- What branches need to be created? Proposed: 3.x and 4.x branches to replace main.
- How will the build process need to change?
- How will the versioning/release process need to change?

## Deliverables
- [x] Create `3.x` branch in GitHub (based on `main`). Set up protected branch settings.
- [x] Update build process for `3.x` branch to allow it to be set as the default branch in GitHub.
- [x] Set `3.x` branch as the default branch in GitHub.
- [x] Update 3.x documentation.
- [x] Create `4.x` branch in GitHub (based on `main`). Set up protected branch settings.
- [x] Clean up `4.x` branch and provision a new application for 4.x development using preferred tools: Vue, Vite, Storybook
- [x] Test out workflow from start to finish. 
- [x] Update 4.x documentation.
- [x] Remove `main` branch.

**Comment by pyrello** [comment]
# Decisions
* 3.x and 4.x branches were created and then immediately set up as protected branches so that all future changes to them will need to be done via PR.
* Protected branch rules:
	* Require a pull request
		* Require approvals
		* Dismiss stale approvals when new commits are pushed
	* Require status checks to pass
		* Require branches to be up-to-date
	* Require linear history
* 4.x branch is being cleaned out to allow starting fresh and to avoid maintaining cruft that wasn't deleted and should have been.
* Initiated a new Vue app in the 4.x branch using the following settings:
	* Add TypeScript? - Yes
	* Add JSX Support? - No
	* Add Vue Router for Single Page Application development? - No
	* Add Pinia for state management? - No
	* Add Vitest for Unit Testing? - Yes
	* Add Cypress for End-to-End testing? - Yes
	* Add ESLint for code quality? - Yes
	* Add Prettier for code formatting? - Yes

**Comment by pyrello** [comment]
The `3.x` branch is now set up as the default branch for this repo. All future development on the 3.x version of UIDS should be done against that branch instead of `main`.

**Comment by pyrello** [comment]
I've been looking into the options we have for being able to theme our Storybook instance. It is somewhat limited in that regard. We should be able to get the colors, fonts, and spacing right. But changing the layout of the page is much more complicated and I'm not exactly sure how we can do it.

Based on this, it seems like we might have to accept a layout that is different than what our brand would indicate is preferred.

There is one other potential way we could handle it. It is possible to embed stories from our storybook into another application either via iframe: https://storybook.js.org/docs/react/sharing/embed. This opens up the possibility of being able to embedding directly into brand.uiowa.edu. However, this approach would require manually keeping all relevant pages up-to-date with any information that is not being embedded from our storybook.

Resources related to theming:
* Creating a Storybook theme: https://storybook.js.org/docs/vue/configure/theming
* Modifying how stories render: https://storybook.js.org/docs/vue/configure/story-rendering
* Injecting additional markup into certain areas of the HTML structure: https://storybook.js.org/docs/vue/configure/story-rendering
* Styling and CSS: https://storybook.js.org/docs/vue/configure/styling-and-css
* Taking over the core Storybook UI (written in React) and reimplementing it ourselves: https://github.com/storybookjs/storybook/tree/next/lib/ui

**Comment by pyrello** [comment]
Came across this and dropping it here for future reference: https://github.com/viljamis/vue-design-system

**Comment by pyrello** [comment]
We talked about my [previous comment](https://github.com/uiowa/uids/issues/671#issuecomment-1100907795) and decided that we weren't worried about trying to mess with the layout of storybook for now.

I have worked on the Button and Headline components and built stories for them. I added a link to the feature branch preview of our storybook instance in the issue description for anyone who is interested. Un-assigning myself now because I'll be out for the next two weeks.
